src/dust/coagulation/coagulation.hpp

```diff
@@ -0,0 +1,719 @@
+    re *= std::sqrt(2.0 * M_PI) * hg;
+  }
+
+  const Real tn = 1.0 / omega;
+  const Real ts = tn / sqrt(re);
+  const Real vn = std::sqrt(alpha) * cs;
+  const Real vs = vn * std::pow(re, -0.25);
+
+  // calculate the relative velocities
+  const Real c1 = 8.0 / M_PI * cs * cs * mu * m_p;
+
+  const Real stokes_i = tau_i * omega;
+  const Real stokes_j = tau_j * omega;
+
+  // turbulent relative velocity
+  Real dv0 = v_rel_ormel(tau_i, tau_j, tn, vn, ts, vs, re);
+
+  // Brownian motion relative velocities
+  const Real mredu = (mass_gridi * mass_gridj / (mass_gridi + mass_gridj));
+  dv0 += (c1 / mredu);
+
+  Real dv2_ij;
+  Real hij = 1.0;
+  if (coag.coord == 1) { // surface density
+    const Real hi =
+        std::min(std::sqrt(alpha / (std::min(0.5, stokes_i) * (SQR(stokes_i) + 1.0))),
+                 1.0) *
+        hg;
+    const Real hj =
+        std::min(std::sqrt(alpha / (std::min(0.5, stokes_j) * (SQR(stokes_j) + 1.0))),
+                 1.0) *
+        hg;
+    const Real vs_i = std::min(stokes_i, 1.0) * omega * hi;
+    const Real vs_j = std::min(stokes_j, 1.0) * omega * hj;
+    // relative velocity from vertical settling
+    dv2_ij = SQR(vs_i - vs_j);
+    dv2_ij += (SQR(vel_i[0] - vel_j[0]) + SQR(vel_i[1] - vel_j[1]));
+    hij = std::sqrt(2.0 * M_PI * (SQR(hi) + SQR(hj)));
+  } else {
+    dv2_ij =
+        (SQR(vel_i[0] - vel_j[0]) + SQR(vel_i[1] - vel_j[1]) + SQR(vel_i[2] - vel_j[2]));
+  }
+
+  // after adding up all v_rel**2 take the square root
+  const Real dv = std::sqrt(dv0 + dv2_ij);
+
+  // new pf calculation: for fragmenation
+  Real pf = 0.0;
+  if (dv > 0.0) {
+    const Real tmp = 1.5 * SQR(coag.vfrag / dv);
+    pf = (tmp + 1.0) * std::exp(-tmp);
+  }
+
+  int icoef_fett = coag2DRv::coef_fett;
+  const Real &coef_fettij = coag.coagR3D(icoef_fett, i, j);
+  if (itype == 0) {                 // coagulation
+    if (coag.ibounce && dv > 0.0) { // including bouncing effect
+      const Real froll = 1e-4;      // Heim et al.(PRL) 1999
+      const Real amono = 1e-4;      // micro-size
+      const Real vbounce = std::sqrt(5.0 * M_PI * amono * froll / mredu);
+      if (vbounce < coag.vfrag) {
+        const Real tmp = 1.5 * SQR(vbounce / dv);
+        pf = (tmp + 1.0) * std::exp(-tmp); // using bouncing vel
+      }
+    }
+    const Real pc = 1.0 - pf;
+    return (coef_fettij * dv * pc / hij);
+  } else { // fragmentation
+    return (coef_fettij * dv * pf / hij);
+  }
+}
+
+KOKKOS_INLINE_FUNCTION
+void CoagulationSource(parthenon::team_mbr_t const &mbr, ScratchPad1D<Real> &source,
+                       const ScratchPad1D<Real> &distri, const int mimax,
+                       const Real kernel4[], const ScratchPad1D<Real> &vel,
+                       const ScratchPad1D<Real> &stoppingTime, const CoagParams &coag) {
+
+  const int nm = coag.nm;
+
+  // initialize source(*)
+  parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, mbr, 0, nm - 1, [&](const int k) {
+    // for (int k = 0; k < nm; k++) {
```

Comment:
can this be deleted?

EDIT: I am actually seeing a lot of commented out for loops in this file.